File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/job/meta/PrepareJobDownloadRequest.scala
package com.webank.eggroll.core.deepspeed.job.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.containers.meta.ContentType
import com.webank.eggroll.core.meta.Deepspeed
import com.webank.eggroll.core.meta.DeepspeedDownload.PrepareDownloadRequest

import scala.collection.JavaConverters.{asJavaIterableConverter, asScalaBufferConverter}
import scala.language.implicitConversions

case class PrepareJobDownloadRequest(sessionId: String, ranks: Array[Int],
                                     compressMethod: String,
                                     compressLevel: Int = 1,
                                     contentType: ContentType.ContentType)


object PrepareJobDownloadRequest {
  implicit def serialize(src: PrepareJobDownloadRequest): Array[Byte] = {
    PrepareDownloadRequest.newBuilder()
      .setSessionId(src.sessionId)
      .addAllRanks(src.ranks.map(_.asInstanceOf[java.lang.Integer]).toSeq.asJava)
      .setCompressMethod(src.compressMethod)
      .setCompressLevel(src.compressLevel)
      .build()
      .toByteArray
  }

  implicit def deserialize(byteString: ByteString): PrepareJobDownloadRequest = {
    val proto = PrepareDownloadRequest.parseFrom(byteString)
    PrepareJobDownloadRequest(
      sessionId = proto.getSessionId,
      ranks = proto.getRanksList.asScala.toArray.map(_.toInt),
      compressMethod = proto.getCompressMethod,
      compressLevel = proto.getCompressLevel,
      contentType = proto.getContentType
    )
  }
}

File: jvm/core/src/main/java/org/fedai/eggroll/core/pojo/PrepareJobDownloadRequest.java
package org.fedai.eggroll.core.pojo;

import com.google.protobuf.InvalidProtocolBufferException;
import com.webank.eggroll.core.meta.DeepspeedDownload;
import lombok.Data;
import org.apache.commons.lang3.StringUtils;

import java.util.List;

@Data
public class PrepareJobDownloadRequest implements RpcMessage {

    String sessionId;
    List<Integer> ranks;
    String compressMethod;
    Integer compressLevel = 1;
    String contentType;

    @Override
    public byte[] serialize() {
        DeepspeedDownload.PrepareDownloadRequest.Builder builder = DeepspeedDownload.PrepareDownloadRequest.newBuilder();
        builder.setSessionId(sessionId);
        if (ranks != null) {
            builder.addAllRanks(ranks);
        }
        if (StringUtils.isNotEmpty(compressMethod)) {
            builder.setCompressMethod(compressMethod);
        }
        builder.setCompressLevel(compressLevel);
        return builder.build().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            DeepspeedDownload.PrepareDownloadRequest request = DeepspeedDownload.PrepareDownloadRequest.parseFrom(data);
            this.sessionId = request.getSessionId();
            this.ranks = request.getRanksList();
            this.compressMethod = request.getCompressMethod();
            if (request.getCompressLevel() > 0) {
                compressLevel = request.getCompressLevel();
            }
        } catch (InvalidProtocolBufferException e) {
            e.printStackTrace();
        }
    }
}
